### PR TITLE
fix/memory leaks and order book desynchronization

### LINKED ISSUE #4683: Gate_io: High memory usage
**Describe the bug**
Opening ticket for tracking: We have observe a high memory usage on our gate_io test bot while testing a pull request. In the UI, the memory stays within 200+MB however when we check the memory usage using HTOP, it increases over time. The highest so far that we've seen reached up to 3+GB. This behaviour persist as well in the development branch.

**Steps To Reproduce**

Here are the steps to reproduce the issue (see attachments in the section below):
1. Create an Avellaneda MM, Liquidity mining, or pure market making strategy using Gate_io
2. Use ETH_USDT as market
3. Start the strategy and leave it running for long period.
4. Monitor the memory usage from time to time using top or htop
...

**Screenshots**
After running the bot for about 24 hours or more:
image

After running the bot for 6 hours:
image

**Release version**
0.45.0

**Attachments**

*WARNING: Do NOT publish any exchange API keys or your wallet's private key. Whoever has access to them may steal your assets!*

**Optional: your discord username:**

### PR DESCRIPTION
**Pull request by aarmoa**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
When analyzing memory consumption for #4683 it was found that the EventLogger keeps an array in memory with all events triggered since the client is started.
This PR reduces that storage to the last 50 events. The only exception are the order filled events. The client needs to keep all of them because they are used to calculate PnL and update balance calculation. We might need an improvement in the future to make it possible to do those calculations without requiring all the events.

To facilitate testing I have added the physical memory consumption to the process monitor indicators in the lower right corner of the screen:

image

UPDATE:
I discovered a memory leak in the order book data source from GateIO connector. I include the fix as part of the PR. QA team please test that the order book is updated correctly after the fix.

UPDATE 2:
During the testing phase it was discovered that the connector was not keeping the local order book synchronized with the server order book all the time.
I include some changes to start using the numeric id included in order book updates messages and snapshot messages as the identifier, instead of the timestamp.

**Tests performed by the developer**:
Tested by running pure market making strategy for ETH-USDT connected to GateIO exchange.

**Tips for QA testing**:
- Verify the order book is updated correctly.

Fixes #4683

### CODE REVIEW

**Review by CrimsonJacket — APPROVED** [review]
Changes LGTM

File: test/hummingbot/connector/exchange/ndax/test_ndax_exchange.py [hunk 1]
```diff
@@ -17,12 +17,13 @@
...
 from hummingbot.connector.exchange.ndax.ndax_order_book import NdaxOrderBook
 from hummingbot.connector.trading_rule import TradingRule
+from hummingbot.core.event.event_logger import EventLogger
 from hummingbot.core.event.events import (
     MarketOrderFailureEvent,
     OrderCancelledEvent,
     OrderFilledEvent,
     OrderType,
-    TradeType,
+    TradeType, MarketEvent,
```
[the remark this answers is not in the document]
Comment (aarmoa) [reply test/hummingbot/connector/exchange/ndax/test_ndax_exchange.py:26]: Good catch

**Review by CrimsonJacket — DISMISSED** [review]
Changes LGTM

**Comment by aarmoa** [comment]
I have found one more issue causing memory leaks, in this case directly related to GateIO connector implementation. I will work on fixing. I am moving the task back to `In Progress` and I will mark the PR as draft momentarily.

**Comment by rxlxrxsx** [comment]
- On less liquid markets there are times that the bid/ask volume and best bid/ask is not in sync

https://user-images.githubusercontent.com/64377055/142819722-f0ea0166-b88c-4fa8-941d-da85ffd4f4b4.mp4

https://user-images.githubusercontent.com/64377055/142819751-66ad270b-ca03-425b-a13d-c2f638801d89.mp4

**Review by CrimsonJacket — APPROVED** [review]
Changes LGTM.

**Comment by aarmoa** [comment]
I have committed some changes to solve the order book synchronization issue.

**Review by CrimsonJacket — APPROVED** [review]
Changes LGTM

**Comment by rxlxrxsx** [comment]
Upon checking again the order book is still not in sync with the client.

https://user-images.githubusercontent.com/64377055/142967824-47da6f8d-0937-4b81-9d33-2e32e5ccb3bc.mp4

https://user-images.githubusercontent.com/64377055/142967841-934ba6ab-3c9d-4201-a856-8775103538c0.mp4

**Comment by aarmoa** [comment]
Hello @rxlxrxsx. According to your logs, do you know if the order book desynchronization happens near some connections error between the client and the exchange?

**Comment by rxlxrxsx** [comment]
Hi @aarmoa, checking the log file there were no connection errors during order book desynchronization. Here's the log file:
[logs_gateio_mem-leak.log.zip](https://github.com/CoinAlpha/hummingbot/files/7591759/logs_gateio_mem-leak.log.zip)

**Comment by aarmoa** [comment]
I have committed a new set of changes. I have enlarged the buffer of diff messages to keep for re-apply if a new snapshot of the order book is taken (it was 32 and it is now 100 for GateIO). I have also updated the comparison logic for GateIO order book messages, to consider the use of the update id instead of only using the timestamp.
@rxlxrxsx and @RC-13 in my local tests, after adding more information in the logs, I could confirm that we are not missing any diff message. The client processes all messages received from the exchange in sequence and there are no gaps.

**Comment by aarmoa** [comment]
I have created a new branch (`fix/memory_leaks_extra_logs`) with a change to log all diff messages and snapshot messages that the client receives from AscendEx.
Each diff message will appear in the logs like this:
```
2021-11-25 00:35:30,829 - 63243 - hummingbot.connector.exchange.gate_io.gate_io_order_book_tracker - INFO - 
### diff message id 28186658 - snapshot_uid 28186657
        GateIoOrderBookMessage(type=<OrderBookMessageType.DIFF: 2>, content={'t': 1637811329680, 'e': 'depthUpdate', 'E': 1637811329, 's': 'KAI_USDT', 'U': 28186658, 'u': 28186658, 'b': [['0.13998', '7038.960390435']], 'a': [], 'trading_pair': 'KAI-USDT', 'update_id': 28186658}, timestamp=1637811329.68)
```

Each snapshot message will appear in the logs like this:
```
2021-11-25 00:35:30,830 - 63243 - hummingbot.connector.exchange.gate_io.gate_io_order_book_tracker - INFO - 
@@@ snapshot message id 28186657 - current ob snapshot id 28186657 - last diff id 0
        GateIoOrderBookMessage(type=<OrderBookMessageType.SNAPSHOT: 1>, content={'id': 28186657, 'current': 1637811329618, 'update': 1637811328037, 'asks': [['0.13999', '3847.68'], ['0.14', '9459.20330493'], ['0.14033', '99.77'], ['0.1404', '7.35'], ['0.14047', '8.81'], ['0.14056', '9241.79'], ['0.14068', '40.3'], ['0.14088', '283.79'], ['0.14099', '20342.69'], ['0.14115', '6244.05']], 'bids': [['0.13998', '5253.000390435'], ['0.13997', '20143.82766337'], ['0.1393', '717.87'], ['0.13919', '532.443816065'], ['0.13909', '13238.64'], ['0.13901', '273.36'], ['0.13898', '15295.88'], ['0.13862', '40.3'], ['0.13856', '15609.75'], ['0.13814', '8085.36']], 'trading_pair': 'KAI-USDT', 'update_id': 28186657}, timestamp=1637811329)
```

The logs will enable QA to determine if the client receives or not an update that is shown in AscendEx web page.

File: hummingbot/connector/exchange/gate_io/gate_io_active_order_tracker.pyx [hunk 2]
```diff
@@ -99,8 +99,8 @@ cdef class GateIoActiveOrderTracker:
 
     cdef tuple c_convert_snapshot_message_to_np_arrays(self, object message):
         cdef:
-            float price
-            float amount
+            object price
+            double amount
```
Comment (CrimsonJacket) [inline hummingbot/connector/exchange/gate_io/gate_io_active_order_tracker.pyx:103]: Is there a reason for this change?

**Review by CrimsonJacket — CHANGES_REQUESTED** [review]
I believe an unintentional change was included. Other than that, LGTM.

File: hummingbot/connector/exchange/gate_io/gate_io_active_order_tracker.pyx (hunk 2 above)
Comment (aarmoa) [reply hummingbot/connector/exchange/gate_io/gate_io_active_order_tracker.pyx:103]: This change was required to keep the same price values for the diff messages and for the snapshot messages. In lines 63 and 66 of the same file you can notice that those are the types used to process prices and amounts for the diff messages. Without using those same types the snapshots processor was converting prices to floats and adding more decimals. That caused that when an update arrived for that price in a diff message, prices did not match, and the original entry in the OrderBook was not changed.

**Comment by aarmoa** [comment]
@rxlxrxsx if possible please use the "Professional" GateIO interface to compare the order book. If you compare the standard and the professional views, even the order books in their servers have differences.
image

**Comment by PtrckM** [comment]
- standard vs profession on client
  - most pairs are sync either use standard or pro on client
  - some pairs are matched either using standard or pro (one of the 2 are not matched) on client
  - also found out either pro or standard, there are still pairs that are not matching the order book example AXS-USDT

image
- standard vs pro vs client = mismatched

https://user-images.githubusercontent.com/54516858/143510973-11f0cd4b-20ce-4f63-ac13-cb244a015bb3.mov

- logs
[logs_gate_pure_mm_avax-usdt.log](https://github.com/CoinAlpha/hummingbot/files/7605632/logs_gate_pure_mm_avax-usdt.log)

**Comment by RC-13** [comment]
For the Gate.io order_book depth, the bot can handle up to 44 lines which is good enough.

**Review by CrimsonJacket — APPROVED** [review]
Changes LGTM 🐦  🔉

**Comment by aarmoa** [comment]
I am merging this PR based on latest discussion with QA. The order books seems to be synchronizing after a couple of seconds every time, and that is acceptable (we depend on the exchange sending the information to update the client order book)

**Review by RC-13 — APPROVED** [review]
LGTM, for the order book synchronization, we all agreed to merge what we currently have and look back if we have many users that will be reporting a desynchronization problem.

**Pull request merged**